### PR TITLE
Task6

### PR DESCRIPTION
**Pull request by alesyalis**

### CODE REVIEW

File: ClassLibrary1/ClassLibrary1/Class1.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ClassLibrary1
+{
+    public class Class1
```
Comment (denja2) [inline ClassLibrary1/ClassLibrary1/Class1.cs]: ???

File: Weather.СonsoleApp/Weather.BL/DTOs/ForecastResponseMessage.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Weather.BL.DTOs
+{
+    public class ForecastResponseMessage
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/DTOs/ForecastResponseMessage.cs]: Don't create duplicate

File: Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs [hunk 3]
```diff
@@ -15,26 +16,44 @@ public class WeatherService : IWeatherService
...
 
             var description = GetWeatherDescription(weather);
-            
-            return MapToWeatherResponseDTO(weather, description);
 
+            return GetWeatherResponseMessage(weather, description);
+
+        }
+
+        public async Task<List<ForecastResponseMessage>> GetForecastAsync(string cityName, int days)
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs]: Add input class

File: Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs [hunk 4]
```diff
@@ -15,26 +16,44 @@ public class WeatherService : IWeatherService
...
+        {
+            _validator.ValidateCityByName(cityName);
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs]: ValidateForecastRequest

File: Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs (hunk 3 above)
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs]: return WeatherResponseMessage

File: Weather.СonsoleApp/Weather.DataAccess/Models/ForecastDescription.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Weather.DataAccess.Models
+{
+    public class ForecastDescription : TemperatureInfo
+    {
+        public double Temp_min { get; set; }
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.DataAccess/Models/ForecastDescription.cs]: Use attributes

File: Weather.СonsoleApp/Weather.DataAccess/Repositories/WeatherRepository.cs [hunk 6]
```diff
@@ -21,12 +23,38 @@ public WeatherRepository(IConfig configuration)
...
             var key = _configuration.ApiKey;
-            var responce = await _httpClient.GetAsync($"{api}={cityName}&units=metric&appid={key}");
+            var responce = await _httpClient.GetAsync($"{urlBase}{urlWeather}={cityName}&units=metric&appid={key}");
             var responceBody = await responce.Content.ReadAsStringAsync();  
             var weatherResponсe = JsonConvert.DeserializeObject<WeatherResponse>(responceBody);
             return weatherResponсe;
         }
+        public async Task<ForecastResponse> GetForecastAsync(string cityName, int days)
+        {
+            var location = await GetLocationAsync(cityName);
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.DataAccess/Repositories/WeatherRepository.cs]: locations

File: Weather.СonsoleApp/Weather.DataAccess/Repositories/WeatherRepository.cs [hunk 7]
```diff
@@ -21,12 +23,38 @@ public WeatherRepository(IConfig configuration)
...
+            var lat = location?.FirstOrDefault()?.Lat;
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.DataAccess/Repositories/WeatherRepository.cs]: Get first location? then validate for null

File: Weather.СonsoleApp/Weather.DataAccess/Repositories/WeatherRepository.cs [hunk 8]
```diff
@@ -21,12 +23,38 @@ public WeatherRepository(IConfig configuration)
...
+            var lon = location?.FirstOrDefault()?.Lon;
+            var urlBase = _configuration.UrlBase;
+            var urlForecast = _configuration.UrlForecast;
+            var key = _configuration.ApiKey;
+
+            var responce = await _httpClient.GetAsync($"{urlBase}{urlForecast}lat={lat}&lon={lon}&cnt={days}&units=metric&appid={key}");
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.DataAccess/Repositories/WeatherRepository.cs]: response

**Review by denja2 — CHANGES_REQUESTED** [review]

File: Weather.СonsoleApp/Weather.BL/Validators/Abstract/Validator.cs [hunk 9]
```diff
@@ -11,5 +11,15 @@ public void ValidateCityByName(string name)
                 throw new ValidationException("Entering the city is required\n");
             }
         }
+
+        public void ValidateForecast(string name, int days)
+        {
+            ValidateCityByName(name);
+
+            if (days == 0)
```
Comment (Xodoc) [inline Weather.СonsoleApp/Weather.BL/Validators/Abstract/Validator.cs]: Why is it only checked for 0 ? U need to consider the option with a negative value

File: Weather.СonsoleApp/Weather.IntegrationTest/appsettings.json [hunk 10]
```diff
@@ -1,6 +1,9 @@
 {
   "ApiKey": "8e943ed8b016561c73b8a1920366ef79",
-  "Url": "https://api.openweathermap.org/data/2.5/weather?q"
-    
-  
+  "Url": "https://api.openweathermap.org/data/2.5/weather?q",
```
Comment (Xodoc) [inline Weather.СonsoleApp/Weather.IntegrationTest/appsettings.json:3]: Remove "weather?q", u have urlWeather

File: Weather.СonsoleApp/Weather.СonsoleApp/Program.cs [hunk 11]
```diff
@@ -24,47 +26,37 @@ static async Task Main(string[] args)
...
-        }
+            ICommand getWeather = new GetWeatherCommand(_weatherService);
+            ICommand getForecast = new GetForecastCommand(_weatherService);
 
-        private static async Task<bool> MainMenu()
-        {
-            Console.WriteLine("Select an action:");
-            Console.WriteLine("If you want to know the weather in the city, enter 1");
-            Console.WriteLine("If you want to close the application, enter 2");
+            var listCommand = new List<ICommand>();
```
Comment (Xodoc) [inline Weather.СonsoleApp/Weather.СonsoleApp/Program.cs]: var listCommand = new List<ICommand>{ getWeather, getForecast };
I think it will be better 🙂

**Review by Xodoc — APPROVED** [review]

File: Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs [hunk 12]
```diff
@@ -15,47 +15,87 @@ public class WeatherService : IWeatherService
...
-            var description = weatherResponse.Main.Description; 
+            var description = weatherResponse.Main.Description;
+            return GetDescription(temperature);
+        }
+
+        private string GetForecastDescription(InfoForecast infoForecast)
+        {
+            var forecastDescription = infoForecast.Main;
+            var temperature = forecastDescription.Temp;
+            var description = forecastDescription.Description;
```
Comment (ihhyy) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs]: Do you need description here ?

File: Weather.СonsoleApp/Weather.IntegrationTest/Service/WeatherServiceIntegrationTest.cs [hunk 13]
```diff
@@ -38,12 +45,55 @@ public async Task GetWeatherAsync_CorrectWeatherReceived_IsErrorFalseAndMessageI
...
+            }
+
+            //Act
+            var response = await _weatherService.GetForecastAsync(name, days);
+
+            // Assert
+            Assert.False(response.IsError);
+            Assert.Matches(message, response.Message);
+        }
+
```
Comment (ihhyy) [inline Weather.СonsoleApp/Weather.IntegrationTest/Service/WeatherServiceIntegrationTest.cs:88]: Spaces

File: Weather.СonsoleApp/Weather.IntegrationTest/Service/WeatherServiceIntegrationTest.cs [hunk 14]
```diff
@@ -71,5 +121,49 @@ public async Task GetWeatherAsync_ValidatorThrowsIsExeption_ReceivedError()
...
+        public async Task GetForecastsync_ReceivedIncorrectWeather_IsErrorTrue()
+        {
+            // Arrange
+            var name = "qqqqqq";
+            var days = 2;
+            var message = $"{name} not found";
+
+            //Act
+            var response = await _weatherService.GetForecastAsync(name, days);
+
```
Comment (ihhyy) [inline Weather.СonsoleApp/Weather.IntegrationTest/Service/WeatherServiceIntegrationTest.cs:127]: Space

**Review by ihhyy — APPROVED** [review]

File: Weather.СonsoleApp/AppConfiguration/AppConfig/Config.cs [hunk 15]
```diff
@@ -4,8 +4,16 @@ namespace AppConfiguration.AppConfig
 {
     public class Config : IConfig
     {
-        public string Url { get { return ConfigurationManager.AppSettings["url"]; } }
+        public string UrlBase { get { return ConfigurationManager.AppSettings["urlBase"]; } }
```
Comment (denja2) [inline Weather.СonsoleApp/AppConfiguration/AppConfig/Config.cs]: I missed it previously, not the best way to initialize config

File: Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs [hunk 16]
```diff
@@ -15,47 +15,86 @@ public class WeatherService : IWeatherService
...
                 Message = $"In {weatherResponse.Name}: {weatherResponse.Main.Temp} °C now. {description} "
             };
             return weatherDTO;
         }
 
+        private ResponseMessage GetForecastMessage(ForecastResponse forecastResponse)
+        {
+            var responseMessage = new ResponseMessage { };
+           
+            forecastResponse.List.ForEach(x => responseMessage.Message += string.Join(",", $"{x.Date.DayOfWeek} In {forecastResponse.CityName.Name}" +
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs]: Create first part of message and then second created by join

File: Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs (hunk 12 above)
Comment (denja2) [reply Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs]: +

File: Weather.СonsoleApp/Weather.IntegrationTest/ConfigTest.cs [hunk 17]
```diff
@@ -5,12 +5,20 @@ namespace Weather.IntegrationTest
...
                 return new ConfigurationBuilder().AddJsonFile("appsettings.json")
-                                                 .Build()["url"];
+                                                 .Build()["urlBase"];
+            }
+        }
+        public string UrlWeather
+        {
+            get
+            {
+                return new ConfigurationBuilder().AddJsonFile("appsettings.json")
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.IntegrationTest/ConfigTest.cs]: No need to read config file in each property

File: Weather.СonsoleApp/Weather.IntegrationTest/Service/WeatherServiceIntegrationTest.cs [hunk 18]
```diff
@@ -38,12 +46,57 @@ public async Task GetWeatherAsync_CorrectWeatherReceived_IsErrorFalseAndMessageI
 
             //Act
             var response = await _weatherService.GetWeatherAsync(name);
-            
-           // Assert
+
+            // Assert
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.IntegrationTest/Service/WeatherServiceIntegrationTest.cs:48]: ?

File: Weather.СonsoleApp/Weather.IntegrationTest/Service/WeatherServiceIntegrationTest.cs [hunk 19]
```diff
@@ -71,5 +124,48 @@ public async Task GetWeatherAsync_ValidatorThrowsIsExeption_ReceivedError()
             var result = Assert.Throws<ValidationException>(() => _validator.ValidateCityByName(name));
             Assert.Equal(message, result.Message);
         }
+
+        [Fact]
+        public async Task GetForecastsync_ReceivedIncorrectWeather_IsErrorTrue()
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.IntegrationTest/Service/WeatherServiceIntegrationTest.cs]: check the method name

File: Weather.СonsoleApp/Weather.BL/Validators/Abstract/Validator.cs [hunk 20]
```diff
@@ -11,5 +11,15 @@ public void ValidateCityByName(string name)
                 throw new ValidationException("Entering the city is required\n");
             }
         }
+
+        public void ValidateForecast(string name, int days)
+        {
+            ValidateCityByName(name);
+
+            if (days <= 0)
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Validators/Abstract/Validator.cs]: !!!! days from config

**Review by denja2 — CHANGES_REQUESTED** [review]

**Pull request merged**